### ISSUE TITLE
Chef cookbook fails on ubuntu 12.04

### ISSUE THREAD

**Issue by libanf** [signalfx/signalfx-agent#830]
The chef cookbook fails to run on ubuntu 12.04, error as follows:

```
  ArgumentError
  -------------
  Malformed version number string 0.8.16~exp12ubuntu10.21

  Cookbook Trace:
  ---------------
    /var/chef.cloudreach/cache/cookbooks/signalfx_agent/recipes/default.rb:42:in `block in from_file'
    /var/chef.cloudreach/cache/cookbooks/signalfx_agent/recipes/default.rb:35:in `from_file'

  Relevant File Content:
  ----------------------
  /var/chef.cloudreach/cache/cookbooks/signalfx_agent/recipes/default.rb:

   35:    package 'signalfx-agent' do # ~FC009
   36:      action :install
   37:      version node['signalfx_agent']['package_version'] unless node['signalfx_agent']['package_version'].nil?
   38:      flush_cache [ :before ] if platform_family?('rhel')
   39:      options '--allow-downgrades' if platform_family?('debian') \
   40:        && node['packages'] \
   41:        && node['packages']['apt'] \
   42>>       && Gem::Version.new(node['packages']['apt']['version']) >= Gem::Version.new('1.1.0')
   43:      allow_downgrade true if platform_family?('rhel', 'amazon', 'fedora')
   44:      notifies :restart, 'service[signalfx-agent]', :delayed
   45:    end
   46:  end
   47:
   48:  template node['signalfx_agent']['conf_file_path'] do
   49:    source 'agent.yaml.erb'
   50:    owner node['signalfx_agent']['user']
   51:    group node['signalfx_agent']['group']

  System Info:
  ------------
  chef_version=13.4.24
  platform=ubuntu
  platform_version=12.04
  ruby=ruby 2.4.2p198 (2017-09-14 revision 59899) [x86_64-linux]
  program_name=chef-client worker: ppid=1461;start=22:54:54;
  executable=/opt/chef/bin/chef-client


  Running handlers:
[2019-05-22T22:55:13+00:00] ERROR: Running exception handlers
  Running handlers complete
[2019-05-22T22:55:13+00:00] ERROR: Exception handlers complete
  Chef Client failed. 2 resources updated in 18 seconds
[2019-05-22T22:55:13+00:00] FATAL: Stacktrace dumped to /var/chef.cloudreach/cache/chef-stacktrace.out
[2019-05-22T22:55:13+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2019-05-22T22:55:13+00:00] ERROR: Malformed version number string 0.8.16~exp12ubuntu10.21
[2019-05-22T22:55:13+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by keitwb** [comment]
Hi @libanf, we don't really support Ubuntu 12.04 anymore.  If I understand it correctly, 12.04 is no longer supported by Canonical either.

The fundamental issue though seems to be that you are using a version of apt that has a very strange version number that the Ruby version module can't understand.  I don't really see any simple workaround for it as I don't know of any way to capture exceptions in Ruby expressions.  @jchengsfx do you know any way around this?

**Comment by jchengsfx** [comment]
A workaround could be to make this change in https://github.com/signalfx/signalfx-agent/blob/master/deployments/chef/recipes/default.rb since `Gem::Version` doesn't like the tilde in the apt package version for 12.04:
```
diff --git a/deployments/chef/recipes/default.rb b/deployments/chef/recipes/default.rb
index 48ea3f15..778f7aa3 100644
--- a/deployments/chef/recipes/default.rb
+++ b/deployments/chef/recipes/default.rb
@@ -69,7 +69,7 @@ else
     options '--allow-downgrades' if platform_family?('debian') \
       && node['packages'] \
       && node['packages']['apt'] \
-      && Gem::Version.new(node['packages']['apt']['version']) >= Gem::Version.new('1.1.0')
+      && Gem::Version.new(node['packages']['apt']['version'].split('~')[0]) >= Gem::Version.new('1.1.0')
     allow_downgrade true if platform_family?('rhel', 'amazon', 'fedora')
     notifies :restart, 'service[signalfx-agent]', :delayed
   end
```

I ran a quick test with this change on 12.04 with chef-client 13.4.24, and was able to install the agent. I'll run more tests to make sure it doesn't introduce any regression.

**Comment by jchengsfx** [comment]
@libanf We released cookbook v0.4.1 to the supermarket with the change above.  Please let us know if that resolves the issues for you.
